woocommerce/woocommerce
# Fix incorrect product data displayed in Product Collection in context of Single Product block

The initial approach is more like POC. Unfortunately, it has an impact on the performance so won't be a way.

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Product Collection (or more technically Product Template) has been using `BlockContextProvider` to pass the product context to nested blocks. This was following the core's Query Loop block.

However, in Blocks world, many blocks are using `ProductDataContextProvider` along with `withProduct` HOC. That includes Single Product block and Product Elements (excluding Product Title and Product Summary that are variations of core blocks).

**Problem:**
The problem was that Product Collection displayed content from Single Product block when used in Single Product block (check Before/After below). This is due to:
- most of Product Elements (e.g. Image, Price) read the `ProductDataContextProvider` set by Single Product parent and wouldn't care about `BlockContextProvider`
- Title read the `BlockContextProvider` set by Product Template and wouldn't care about `ProductDataContextProvider`

**Solution:**
Considering we have inner blocks reacting on two separate contexts we need to pass both of them.

Closes https://github.com/woocommerce/woocommerce/issues/43991



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create new post
2. Add Product Collection block
3. Choose any collection
4. Expected: Make sure all the products display correct data. Example of incorrect data:
  - "Title" as product name
  - Duplicated images
  - Duplicated prices
5. Focus on different product elements (image, price, title) and make sure there's no blinking or images reloading
6. Add Single Product block
7. Add Product Collection block inside of Single Product block _(note: make sure you're adding it inside some **Column** block. It's forbidden to add any block inside **Columns** block)_
8. Expected: Make sure all the products display correct data. Example of incorrect data:
  - "Title" as product name
  - Duplicated images or images taken from wrapping Single Product block
  - Duplicated prices or prices taken from wrapping Single Product block
9. Focus on different product elements (image, price, title) and make sure there's no blinking or images reloading



Before | After
-- | --
<img width="1404" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/9bc4e1de-de0e-45fa-af32-c6e2696f3d47"> | <img width="1196" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/22be0f1f-6c24-4f9f-98d1-19755fbf8223">


### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
